uaclient/data_types.py

```diff
@@ -0,0 +1,133 @@
+from typing import Any, List, Optional, Type, TypeVar
+
+from uaclient.exceptions import UserFacingError
+
+INCORRECT_TYPE_ERROR_MESSAGE = (
+    "Expected value with type {type} but got value: {value}"
+)
+INCORRECT_LIST_ELEMENT_TYPE_ERROR_MESSAGE = (
+    "Got value with incorrect type at index {index}: {nested_msg}"
+)
+INCORRECT_FIELD_TYPE_ERROR_MESSAGE = (
+    'Got value with incorrect type for field "{key}": {nested_msg}'
+)
+
+
+class IncorrectTypeError(UserFacingError):
+    def __init__(self, expected_type: str, got_value: Any):
+        super().__init__(
+            INCORRECT_TYPE_ERROR_MESSAGE.format(
+                type=expected_type, value=repr(got_value)
+            )
+        )
+
+
+class IncorrectListElementTypeError(IncorrectTypeError):
+    def __init__(self, err: IncorrectTypeError, at_index: int):
+        self.msg = INCORRECT_LIST_ELEMENT_TYPE_ERROR_MESSAGE.format(
+            index=at_index, nested_msg=err.msg
+        )
+
+
+class IncorrectFieldTypeError(IncorrectTypeError):
+    def __init__(self, err: IncorrectTypeError, key: str):
+        self.msg = INCORRECT_FIELD_TYPE_ERROR_MESSAGE.format(
+            key=key, nested_msg=err.msg
+        )
+
+
+class DataValue:
+    @staticmethod
+    def from_value(val: Any) -> Any:
+        return val
+
+
+class StringDataValue(DataValue):
+    @staticmethod
+    def from_value(val: Any) -> str:
+        if not isinstance(val, str):
+            raise IncorrectTypeError("string", val)
+        return val
+
+
+class IntDataValue(DataValue):
+    @staticmethod
+    def from_value(val: Any) -> int:
+        if not isinstance(val, int):
+            raise IncorrectTypeError("int", val)
+        return val
+
+
+def data_list(data_cls: Type[DataValue]) -> Type[DataValue]:
+    class _DataList(DataValue):
+        @staticmethod
+        def from_value(val: Any) -> List:
+            if not isinstance(val, list):
+                raise IncorrectTypeError("list", val)
+            ret_list = []
```

Comment:
Why do we need the auxiliary list here ? Can't we just iterate on the original items of the list and perform the validation on then ?